### PR TITLE
Update A&A Forum Poster

### PR DESCRIPTION
**Pull request by ron-murhammer**
Took a quick swing at this but ended up getting this error:
```
SEVERE: Failed to post game to forum
java.lang.IllegalStateException: Failed to retrieve Token, server did not return correct JSON: {"path":"&#x2F;forums&#x2F;api&#x2F;v2&#x2F;users&#x2F;8592&#x2F;tokens","error":"Cannot read property 'uid' of undefined"}
	at games.strategy.engine.pbem.NodeBbForumPoster.getToken(NodeBbForumPoster.java:153)
	at games.strategy.engine.pbem.NodeBbForumPoster.postTurnSummary(NodeBbForumPoster.java:47)
	at games.strategy.engine.pbem.AxisAndAlliesForumPoster.postTurnSummary(AxisAndAlliesForumPoster.java:1)
	at games.strategy.engine.framework.startup.ui.editors.ForumPosterEditor.lambda$3(ForumPosterEditor.java:144)
	at java.lang.Thread.run(Thread.java:748)
```

### CODE REVIEW

File: game-core/src/main/java/games/strategy/engine/pbem/AxisAndAlliesForumPoster.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
+package games.strategy.engine.pbem;
+
+import games.strategy.triplea.UrlConstants;
+import games.strategy.triplea.help.HelpSupport;
+
+/**
+ * Posts turn summaries to forums.triplea-game.org.
+ *
+ * <p>
+ * URL format is {@code https://forums.triplea-game.org/api/v2/topics/<topicID>}.
```
Comment (RoiEXLab) [inline game-core/src/main/java/games/strategy/engine/pbem/AxisAndAlliesForumPoster.java]: This needs to be updated

File: game-core/src/main/java/games/strategy/triplea/UrlConstants.java [hunk 2]
```diff
@@ -25,7 +25,7 @@
 
   TRIPLEA_FORUM("https://forums.triplea-game.org/"),
 
-  AXIS_AND_ALLIES_FORUM("https://www.axisandallies.org/forums/index.php"),
+  AXIS_AND_ALLIES_FORUM("https://axisandallies.org/forums"),
```
Comment (RoiEXLab) [inline game-core/src/main/java/games/strategy/triplea/UrlConstants.java]: Adding a / at the end of the url might already do the trick.

Comment (RoiEXLab) [reply game-core/src/main/java/games/strategy/triplea/UrlConstants.java]: Not really sure about that anymore, it should work nevertheless.
I'll think about this in the next few hours.

Comment (ron-murhammer) [reply game-core/src/main/java/games/strategy/triplea/UrlConstants.java]: I originally tried with a / and was getting a 302.

**Comment by RoiEXLab** [comment]
@ron-murhammer I think I figured out the issue.
There's a bug in nodebb-plugin-write-api that causes the javascript equivalent of a NullPointerException when the authentication is invalid.
So have you created an account and used the correct username and password?
I'll create an issue at the repository to have this fixed btw.

**Comment by RoiEXLab** [comment]
Nevermind, I see your account there.
Make sure your password is correct, I'll have a further look later this day in case that's not the problem

**Comment by ron-murhammer** [comment]
@RoiEXLab Just tested and get a different error this morning. I get the same problem with my correct or incorrect password:
```
SEVERE: End of input at character 0 of 
com.github.openjson.JSONException: End of input at character 0 of 
	at com.github.openjson.JSONTokener.syntaxError(JSONTokener.java:473)
	at com.github.openjson.JSONTokener.nextValue(JSONTokener.java:114)
	at com.github.openjson.JSONObject.init(JSONObject.java:211)
	at com.github.openjson.JSONObject.<init>(JSONObject.java:156)
	at com.github.openjson.JSONObject.<init>(JSONObject.java:168)
	at games.strategy.engine.pbem.NodeBbForumPoster.getToken(NodeBbForumPoster.java:141)
	at games.strategy.engine.pbem.NodeBbForumPoster.postTurnSummary(NodeBbForumPoster.java:47)
	at games.strategy.engine.pbem.AxisAndAlliesForumPoster.postTurnSummary(AxisAndAlliesForumPoster.java:1)
	at games.strategy.engine.framework.startup.ui.editors.ForumPosterEditor.lambda$3(ForumPosterEditor.java:144)
	at java.lang.Thread.run(Thread.java:748)
```

**Comment by ron-murhammer** [comment]
@RoiEXLab You're probably on the right track. I tried correct and incorrect passwords and get that same error.

**Comment by RoiEXLab** [comment]
@ron-murhammer Yeah, figured out the thing with the redirect a minute ago as well.
We might want to configure a redirect strategy for this, so it won't happen in the future.
I recall doing this for the dice server in order to correctly redirect 301 status codes to the HTTPS URL.

**Comment by RoiEXLab** [comment]
@ron-murhammer I found the root cause for the bug and it's definitely not our fault.
Because the URL starts with a "/forum" instead of "/api" this messes up an API check here:
https://github.com/NodeBB/nodebb-plugin-write-api/blob/4eadba6429ccdbd57bc8a0be7bf72833671efe26/routes/v2/middleware.js#L85

Which ultimately makes the write-api mess up the logic.
I'll open an issue/Pull Request for this.

/cc @djensen47

**Comment by RoiEXLab** [comment]
https://github.com/NodeBB/nodebb-plugin-write-api/pull/103

**Comment by RoiEXLab** [comment]
@ron-murhammer
Now that we know this isn't our fault I'd kindly ask you to address the review comments in the meantime.
If your feeling like improving the NodeBbForumPoster code to fix the issues with 301 redirects (have a look at dice server logic) or to be more verbose with status codes in error messages in order to improve the debugging experience in the future (500 error code would have given it away) I'd really appreciate that

**Comment by ron-murhammer** [comment]
@RoiEXLab Yep, nice work tracking that down. I'll try to clean this up a bit later today. I probably won't tackle the redirects as I'm trying to focus most of my energy on #4350 which I don't want to lose context or momentum on.

**Comment by djensen47** [comment]
So the fix is in for the nodebb-plugin-write bug?

**Comment by RoiEXLab** [comment]
Yeah, v5.2.2 will have the fix.
The author is currently having some trouble publishing the new version (I'd assume because of a temporary outage or something), but once the new version is available in the ACP and installed on the server, this should work right out of the box.

**Comment by ron-murhammer** [comment]
@RoiEXLab This should be good now. We can always do a follow up PR to make any other improvements like redirect handling.

**Comment by RoiEXLab** [comment]
@djensen47 Posting it here again for future reference (and because I don't really trust gitter notifications):
The latest nodebb-plugin-write-api version is out, once you updated that everything should work.

**Comment by djensen47** [comment]
It's installed on the server now. Thanks.

File: game-core/src/main/java/games/strategy/engine/pbem/NodeBbForumPoster.java [hunk 3]
```diff
@@ -147,7 +147,8 @@ private String getToken(final CloseableHttpClient client, final int userId) thro
         throw new IllegalStateException(
             "Failed to retrieve Token. Code: " + code + " Message: " + jsonObject.getString("message"));
       }
-      throw new IllegalStateException("Failed to retrieve Token, server did not return correct JSON: " + rawJson);
+      throw new IllegalStateException("Failed to retrieve Token, server did not return correct response: "
+          + response.getStatusLine() + "; JSON: " + rawJson);
```
[the remark this answers is not in the document]
Comment (ron-murhammer) [reply game-core/src/main/java/games/strategy/engine/pbem/NodeBbForumPoster.java]: It prints the code and description of the code.

**Comment by djensen47** [comment]
@RoiEXLab Was that message for me or @ron-murhammer. 

I haven't tested anything.

**Comment by RoiEXLab** [comment]
For @ron-murhammer, no worries

**Comment by ron-murhammer** [comment]
@RoiEXLab Everything seems to work fine except the first time I try to set up a PBF game I get the following error around invalid credentials. Successful post: https://www.axisandallies.org/forums/topic/32423/triplea-test-posts

Error (doesn't seem like a big issue as it only occurs the first time):
```
SEVERE: serialized local bean cache invalid
java.lang.NullPointerException
	at com.google.common.base.Preconditions.checkNotNull(Preconditions.java:882)
	at games.strategy.security.DefaultCredentialManager.unprotectToString(DefaultCredentialManager.java:213)
	at games.strategy.engine.pbem.AbstractForumPoster.unprotectCredentials(AbstractForumPoster.java:87)
	at games.strategy.engine.pbem.AbstractForumPoster.readObject(AbstractForumPoster.java:81)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at java.io.ObjectStreamClass.invokeReadObject(ObjectStreamClass.java:1158)
	at java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:2173)
	at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2064)
	at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1568)
	at java.io.ObjectInputStream.readObject(ObjectInputStream.java:428)
	at java.util.HashMap.readObject(HashMap.java:1409)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at java.io.ObjectStreamClass.invokeReadObject(ObjectStreamClass.java:1158)
	at java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:2173)
	at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2064)
	at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1568)
	at java.io.ObjectInputStream.readObject(ObjectInputStream.java:428)
	at games.strategy.engine.framework.startup.ui.PbemSetupPanel$LocalBeanCache.loadMap(PbemSetupPanel.java:386)
	at games.strategy.engine.framework.startup.ui.PbemSetupPanel$LocalBeanCache.<init>(PbemSetupPanel.java:376)
	at games.strategy.engine.framework.startup.ui.PbemSetupPanel$LocalBeanCache.<clinit>(PbemSetupPanel.java:368)
	at games.strategy.engine.framework.startup.ui.PbemSetupPanel.loadDiceServer(PbemSetupPanel.java:166)
	at games.strategy.engine.framework.startup.ui.PbemSetupPanel.loadAll(PbemSetupPanel.java:152)
	at games.strategy.engine.framework.startup.ui.PbemSetupPanel.<init>(PbemSetupPanel.java:88)
	at games.strategy.engine.framework.startup.mc.SetupPanelModel.showPbem(SetupPanelModel.java:49)
	at games.strategy.engine.framework.startup.ui.MetaSetupPanel.lambda$2(MetaSetupPanel.java:129)
	at javax.swing.AbstractButton.fireActionPerformed(AbstractButton.java:2022)
	at javax.swing.AbstractButton$Handler.actionPerformed(AbstractButton.java:2348)
	at javax.swing.DefaultButtonModel.fireActionPerformed(DefaultButtonModel.java:402)
	at javax.swing.DefaultButtonModel.setPressed(DefaultButtonModel.java:259)
	at javax.swing.plaf.basic.BasicButtonListener.mouseReleased(BasicButtonListener.java:252)
	at org.pushingpixels.substance.internal.utils.RolloverButtonListener.mouseReleased(RolloverButtonListener.java:108)
	at java.awt.AWTEventMulticaster.mouseReleased(AWTEventMulticaster.java:289)
	at java.awt.Component.processMouseEvent(Component.java:6533)
	at javax.swing.JComponent.processMouseEvent(JComponent.java:3324)
	at java.awt.Component.processEvent(Component.java:6298)
	at java.awt.Container.processEvent(Container.java:2237)
	at java.awt.Component.dispatchEventImpl(Component.java:4889)
	at java.awt.Container.dispatchEventImpl(Container.java:2295)
	at java.awt.Component.dispatchEvent(Component.java:4711)
	at java.awt.LightweightDispatcher.retargetMouseEvent(Container.java:4889)
	at java.awt.LightweightDispatcher.processMouseEvent(Container.java:4526)
	at java.awt.LightweightDispatcher.dispatchEvent(Container.java:4467)
	at java.awt.Container.dispatchEventImpl(Container.java:2281)
	at java.awt.Window.dispatchEventImpl(Window.java:2746)
	at java.awt.Component.dispatchEvent(Component.java:4711)
	at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:758)
	at java.awt.EventQueue.access$500(EventQueue.java:97)
	at java.awt.EventQueue$3.run(EventQueue.java:709)
	at java.awt.EventQueue$3.run(EventQueue.java:703)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:80)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:90)
	at java.awt.EventQueue$4.run(EventQueue.java:731)
	at java.awt.EventQueue$4.run(EventQueue.java:729)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:80)
	at java.awt.EventQueue.dispatchEvent(EventQueue.java:728)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:201)
	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:116)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:105)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:93)
	at java.awt.EventDispatchThread.run(EventDispatchThread.java:82)
```

**Comment by RoiEXLab** [comment]
@ron-murhammer This issue is related to the ugly code we have for all the forum posters and mail senders.
Hopefully won't happen in the future when I'm done with the changes.

**Comment by RoiEXLab** [comment]
@ron-murhammer Would you be so kind to create a Pull Request for the 1.9 branch to push this on a new stable?

**Comment by ssoloff** [comment]
> Everything seems to work fine except the first time I try to set up a PBF game I get the following error around invalid credentials.

@ron-murhammer That's due to renaming the `AbstractForumPoster` credential fields in #4314 and then trying to load a _local.cache_ file left over from 1.9.  _local.cache_ will be immediately overwritten with a proper serialized form, which is why you only see the error once.  However, if you go back to 1.9, you'll probably get the same (or a similar) error once again.

**Comment by ssoloff** [comment]
@ron-murhammer To answer your direct question, I believe 1.9 users _would_ see the error if you merged the commits that make up this fix into 1.9.  I should clarify that to say 1.9 users that have played at least one PBF game in the past. Playing such a game would have serialized an `AbstractForumPoster` instance into _local.cache_ at some point, so deserialization will trigger the same error you saw.

Note that even though the error only occurs once, it does have the side effect of erasing all cached PBEM/PBF settings, including credentials.  That's not a show-stopper bug, but it may be considered by some users to be annoying.

Actually, looking at these changes again, I'm not sure you can cherry-pick the commits that make up this fix as-is to the 1.9 branch.  I'm pretty sure some of these types are stored in the save game, as well as _local.cache_, and so will break save game compatibility on the 1.9 branch.  The changes I'm thinking break compatibility include renaming the `AbstractForumPoster` fields (among other classes in the same package whose fields have been renamed), introducing the new `NodeBbForumPoster` superclass, and moving `AxisAndAlliesForumPoster` to a new package.

You may want to test a PBF game using a 1.9 client with the cherry-picked commits and a 12226 client to see if there really is a save game compatibility issue.  If so, I guess that means this change will have to be implemented independently on the 1.9 branch rather than cherry-picked.

**Comment by RoiEXLab** [comment]
@ssoloff I completely forgot about the field Renaming thingy and the compatibility it would break.
Thanks for pointing that out 👍

**Comment by ron-murhammer** [comment]
@ssoloff I see your point. Blah. I'll probably just recreate the changes on 1.9 then and make sure to test it.

**Comment by ron-murhammer** [comment]
@ssoloff And yeah you're correct about save game compatibility and essentially can't touch any of the package/class/field names otherwise you get errors trying to load save games.

**Comment by RoiEXLab** [comment]
It's ugly that the savegame is tied to ths 🤷‍♂️

**Comment by ron-murhammer** [comment]
Indeed. Just has to do with the way PBEM/PBF settings are stored/loaded in save games.

**Comment by simon33-2** [comment]
Just tested this and it seems to work in 1.9.

Main proviso is that the topic ID will need to be changed to the new forum's number. I was thinking it was going to include the thread title in there but fortunately (or unfortunately) it does not.

**Comment by simon33-2** [comment]
Shall we advise the intrepid users over at AxisandAllies.org that they can grab the new version if they want to become beta testers? No doubt many are champing at the bit.

**Comment by ron-murhammer** [comment]
@simon33-2 Yeah, feel free to link the latest 1.9 pre-release to the A&A forum users.

**Pull request merged**